src/stream/src/executor/managed_state/join/mod.rs

```diff
@@ -151,152 +156,184 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
-        if contains {
-            Ok(self.inner.get_mut(key).unwrap())
-        } else {
-            self.init_without_cache(key).await?;
-            Ok(self.inner.get_mut(key).unwrap())
+        side_match: &'a mut Self,
+        keys: impl Iterator<Item = &'a K>,
+    ) -> IterJoinEntriesMut<'a, K, S> {
+        IterJoinEntriesMut::<'a>::new(side_match, self, keys)
+    }
+}
+
+type IterType<K, S> = (K, Option<JoinEntryState<S>>, JoinEntryState<S>);
+
+/// Iterates
+pub(crate) struct IterJoinEntriesMut<'a, K: HashKey, S: StateStore> {
+    side_match: &'a mut JoinHashMap<K, S>,
+    side_update: &'a mut JoinHashMap<K, S>,
+    last_item: Option<IterType<K, S>>,
+    queue: tokio::sync::mpsc::UnboundedReceiver<IterType<K, S>>, /* TODO: change this to a
+                                                                  * reasonably
+                                                                  * sized bounded queue... */
+}
+
+impl<'a, K: HashKey, S: StateStore> IterJoinEntriesMut<'a, K, S> {
+    pub(crate) fn new(
+        side_match: &'a mut JoinHashMap<K, S>,
+        side_update: &'a mut JoinHashMap<K, S>,
+        keys: impl Iterator<Item = &'a K>,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel::<IterType<K, S>>();
+        let new_iter = Self {
+            side_match,
+            side_update,
+            last_item: None,
+            queue: rx,
+        };
+
+        let start = std::time::Instant::now();
+        let mut cache_miss_counter = 0;
+        let mut total_counter = 0;
+
+        for key in keys {
+            total_counter += 1;
+            let update_entry = new_iter.side_update.pop_cached(key);
+            let match_entry = if key.has_null() {
+                None
+            } else {
+                new_iter.side_match.pop_cached(key)
+            };
+            let key = key.clone();
+            let update_table_info = &new_iter.side_update.table_info;
+
+            let update_entry = update_entry.unwrap_or_else(|| {
+                JoinHashMap::<K, S>::init_without_cache(&key, update_table_info)
+                    .ok()
+                    .unwrap()
+            });
+
+            if key.has_null() {
+                tx.send((key, None, update_entry))
+                    .ok()
+                    .expect("Failed to send join entries for join key");
+            } else if match_entry.is_some() {
+                tx.send((key, match_entry, update_entry))
+                    .ok()
+                    .expect("Failed to send join entries for join key");
+            } else {
+                cache_miss_counter += 1;
+                // tx.send((key, None, update_entry))
+                //     .ok()
+                //     .expect("Failed to send join entries for join key");
+                let match_table_info = new_iter.side_match.table_info.clone();
+                let tx = tx.clone();
+                tokio::spawn(async move {
```

Comment:
I think we should have exactly one tokio task in the background. Inside that task, we use `stream::iter`, `select_all` or something like that to ensure that I/O are parallelized while only occupying one tokio task.